runtime/Java/src/org/antlr/v4/runtime/DefaultErrorStrategy.java

```diff
@@ -766,4 +776,98 @@ protected void consumeUntil(Parser recognizer, IntervalSet set) {
             ttype = recognizer.getInputStream().LA(1);
         }
     }
+
+	/**
+	 * Construct a parse tree from the current recognizer position to the end of the rule.
+	 *
+	 * <ul>
+	 * <li>Positive closures match exactly once</li>
+	 * <li>Closures match zero times</li>
+	 * <li>Optional blocks do not match</li>
+	 * <li>Decisions take the first alternative, except when one of the above applies.</li>
+	 * </ul>
+	 *
+	 * @param recognizer The recognizer.
+	 */
+	protected void buildErrorTree(Parser recognizer) {
+		final ATN atn = recognizer.getATN();
+		int state = recognizer.getState();
+		ParserRuleContext context = recognizer.getContext();
+		final IntegerStack returnStateStack = new IntegerStack();
+		while (true) {
+			ATNState p = atn.states.get(state);
+			int edge;
+			switch (p.getStateType()) {
```

Comment:
💭 This should probably handle the decision for a rule:

* If an error occurs in a rule before passing the rule entry decision and the rule has labeled alternatives, we need to make sure the correct context is constructed
* We need to call `context.setAltNumber`

Reply:
➡️ Now fixed